docs/rfcs/component-universal-telemetry.md

```diff
@@ -0,0 +1,118 @@
+3. Define specific metrics for each kind of pipeline component.
+4. Define specific logs for all kinds of pipeline component.
+
+### Mechanism
+
+The mechanism of telemetry capture should be _external_ to components. Specifically, we should observe telemetry at each point where a component passes data to another component, and, at each point where a component consumes data from another component. In terms of the component graph, every _edge_ in the graph will have two layers of instrumentation - one for the producing component and one for the consuming component. Importantly, each layer generates telemetry ascribed to a single component instance, so by having two layers per edge we can describe both sides of each handoff independently.
+
+### Attributes
+
+All signals should use the following attributes:
+
+#### Receivers
+
+- `otel.component.kind`: `receiver`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs`, `metrics`, `traces`, **OR `ANY`**
+
+#### Processors
+
+- `otel.component.kind`: `processor`
+- `otel.component.id`: The component ID
+- `otel.pipeline.id`: The pipeline ID, **OR `ANY`**
+- `otel.signal`: `logs`, `metrics`, `traces`, **OR `ANY`**
+
+#### Exporters
+
+- `otel.component.kind`: `exporter`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs`, `metrics` `traces`, **OR `ANY`**
+
+#### Connectors
+
+- `otel.component.kind`: `connector`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs->logs`, `logs->metrics`, `logs->traces`, `metrics->logs`, `metrics->metrics`, etc, **OR `ANY`**
+
+Notes: The use of `ANY` indicates that values are not associated with a particular signal or pipeline. This is used when a component enforces non-standard instancing patterns. For example, the `otlp` receiver isa singleton, so the values are aggregated across signals. Similarly, the `memory_limiter` processor is a singleton, so the values are aggregated across pipelines.
+
+### Metrics
+
+There are two straightforward measurements that can be made on any pdata:
+
+1. A count of "items" (spans, data points, or log records). These are low cost but broadly useful, so they should be enabled by default.
+2. A measure of size, based on [ProtoMarshaler.Sizer()](https://github.com/open-telemetry/opentelemetry-collector/blob/9907ba50df0d5853c34d2962cf21da42e15a560d/pdata/ptrace/pb.go#L11). These are high cost to compute, so by default they should be disabled (and not calculated).
+
+The location of these measurements can be described in terms of whether the data is "incoming" or "outgoing", from the perspective of the component to which the telemetry is ascribed.
+
+1. Incoming measurements are attributed to the component which is _consuming_ the data.
+2. Outgoing measurements are attributed to the component which is _producing_ the data.
```

Comment:
Challenge for this: Do we really need both? If ever component records this, the outgoing of component X will be equal with next component Y metrics.

Reply:
I was also considering this. Are there situations where `outgoing` metric of component A is not equal to `incoming` metric of component B? I think this is the same information. Does splitting this into two different metrics help in usage?

I was considering a "one metric per graph edge" approach, as described in the below examples.

The advantage of the "one metric per edge" approach is that with sufficiently [elaborate](https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/v4.10.0/deploy/helm/sumologic/conf/logs/otelcol/config.yaml#L562-L591) pipelines, the amount of metrics is cut in half (asymptotically).

Is there information that cannot be expressed with this "one metric per edge" approach? Are there other disadvantages, e.g. usability?

<details><summary>Example 1: Simple pipeline with Filter processor</summary>

```yaml
exporters:
  debug:
processors:
  filter:
    logs:
      log_record:
        - 'severity_number < SEVERITY_NUMBER_WARN'
receivers:
  filelog:
    include:
      - logs.json
service:
  pipelines:
    logs:
      exporters:
        - debug
      processors:
        - filter
      receivers:
        - filelog
```

Metrics according to this proposal:

```console
otelcol_component_incoming_items{otel.component.kind="receiver",otel.component.id="filelog",otel.signal="logs"} 2
otelcol_component_outgoing_items{otel.component.kind="receiver",otel.component.id="filelog",otel.signal="logs"} 2
otelcol_component_incoming_items{otel.component.kind="processor",otel.component.id="filter",otel.signal="logs"} 2
otelcol_component_outgoing_items{otel.component.kind="processor",otel.component.id="filter",otel.signal="logs"} 1
otelcol_component_incoming_items{otel.component.kind="exporter",otel.component.id="exporter",otel.signal="logs"} 1
otelcol_component_outgoing_items{otel.component.kind="exporter",otel.component.id="exporter",otel.signal="logs"} 1
```

Here's another proposal, with one metric per graph edge:

```console
otelcol_component[_incoming]_items{otel[.destination].component.kind="receiver",otel[.destination].component.id="filelog",otel.signal="logs"} 2
otelcol_component_items{otel.source.component.kind="receiver",otel.source.component.id="filelog",otel.destination.component.kind="processor",otel.destination.component.id="filter",otel.signal="logs"} 2
otelcol_component_items{otel.source.component.kind="processor",otel.source.component.id="filter",otel.destination.component.kind="exporter",otel.destination.component.id="debug",otel.signal="logs"} 1
otelcol_component[_outgoing]_items{otel.component.kind="exporter",otel.component.id="exporter",otel.signal="logs"} 1
```
</details>

<details><summary>Example 2: Pipeline with connector</summary>

```yaml
connectors:
  otlpjson:
exporters:
  debug:
receivers:
  filelog:
    include:
      - otlp-log.json
      - otlp-metric.json
      - otlp-span.json
service:
  pipelines:
    logs/input:
      exporters: [otlpjson]
      receivers: [filelog]
    logs/otlp:
      exporters: [debug]
      receivers: [otlpjson]
    metrics/otlp:
      exporters: [debug]
      receivers: [otlpjson]
    traces/otlp:
      exporters: [debug]
      receivers: [otlpjson]
```

This proposal:

```console
otelcol_component_incoming_items{otel.component.kind="receiver",otel.component.id="filelog",otel.signal="logs"} 3
otelcol_component_outgoing_items{otel.component.kind="receiver",otel.component.id="filelog",otel.signal="logs"} 3

otelcol_component_incoming_items{otel.component.kind="connector",otel.component.id="otlpjson",otel.signal="logs->logs"} 3
otelcol_component_incoming_items{otel.component.kind="connector",otel.component.id="otlpjson",otel.signal="logs->metrics"} 3
otelcol_component_incoming_items{otel.component.kind="connector",otel.component.id="otlpjson",otel.signal="logs->traces"} 3
otelcol_component_outgoing_items{otel.component.kind="connector",otel.component.id="otlpjson",otel.signal="logs->logs"} 1
otelcol_component_outgoing_items{otel.component.kind="connector",otel.component.id="otlpjson",otel.signal="logs->metrics"} 1
otelcol_component_outgoing_items{otel.component.kind="connector",otel.component.id="otlpjson",otel.signal="logs->traces"} 1

otelcol_component_incoming_items{otel.component.kind="exporter",otel.component.id="debug",otel.signal="logs"} 1
otelcol_component_outgoing_items{otel.component.kind="exporter",otel.component.id="debug",otel.signal="logs"} 1
otelcol_component_incoming_items{otel.component.kind="exporter",otel.component.id="debug",otel.signal="metrics"} 1
otelcol_component_outgoing_items{otel.component.kind="exporter",otel.component.id="debug",otel.signal="metrics"} 1
otelcol_component_incoming_items{otel.component.kind="exporter",otel.component.id="debug",otel.signal="traces"} 1
otelcol_component_outgoing_items{otel.component.kind="exporter",otel.component.id="debug",otel.signal="traces"} 1
```

One metric per edge:

```console
otelcol_component[_incoming]_items{otel[.destination].component.kind="receiver",otel[.destination].component.id="filelog",otel.signal="logs"} 3

otelcol_component_items{otel.source.component.kind="receiver",otel.source.component.id="filelog",otel.destination.component.kind="connector",otel.destination.component.id="otlpjson",otel.signal="logs"} 3
otelcol_component_items{otel.source.component.kind="receiver",otel.source.component.id="filelog",otel.destination.component.kind="connector",otel.destination.component.id="otlpjson",otel.signal="metrics"} 3
otelcol_component_items{otel.source.component.kind="receiver",otel.source.component.id="filelog",otel.destination.component.kind="connector",otel.destination.component.id="otlpjson",otel.signal="traces"} 3

otelcol_component_items{otel.source.component.kind="connector",otel.source.component.id="otlpjson",otel.destination.component.kind="exporter",otel.destination.component.id="debug",otel.signal="logs"} 1
otelcol_component_items{otel.source.component.kind="connector",otel.source.component.id="otlpjson",otel.destination.component.kind="exporter",otel.destination.component.id="debug",otel.signal="metrics"} 1
otelcol_component_items{otel.source.component.kind="connector",otel.source.component.id="otlpjson",otel.destination.component.kind="exporter",otel.destination.component.id="debug",otel.signal="traces"} 1

otelcol_component[_outgoing]_items{otel[.source].component.kind="exporter",otel[.source].component.id="exporter",otel.signal="logs"} 1
otelcol_component[_outgoing]_items{otel[.source].component.kind="exporter",otel[.source].component.id="exporter",otel.signal="metrics"} 1
otelcol_component[_outgoing]_items{otel[.source].component.kind="exporter",otel[.source].component.id="exporter",otel.signal="traces"} 1
```

</details>

Reply:
> Do we really need both? If ever component records this, the outgoing of component X will be equal with next component Y metrics.

This is only true for linear pipelines. Any fanout or merge of data streams breaks this. e.g. Receiver A emits 10 items, receiver B emits 20, processor C consumes 30.

> The advantage of the "one metric per edge" approach is that with sufficiently [elaborate](https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/v4.10.0/deploy/helm/sumologic/conf/logs/otelcol/config.yaml#L562-L591) pipelines, the amount of metrics is cut in half (asymptotically).

Philosophically, I think users should reasonably expect that they can observe an object in isolation to get a relatively complete picture of how it is behaving. Offloading half of the picture for the sake of reducing data volume is a premature optimization in my opinion. In any case, if we support in [some way the ability to disable individual metrics](https://github.com/open-telemetry/opentelemetry-collector/issues/10769), then users can just disable one or the other to opt into exactly this same optimization.

More concretely, say we measure only incoming values. Some things that immediately become difficult for users:
1. How much data is my filter processor discarding? I can find the incoming value easily, but in order to answer the question I also have to figure out which component(s) it is sending data to, query for _their_ incoming values, aggregate those values, then compare to the incoming. I'm sure this is easier in some backends than others but in I don't think users will often write queries for this - they will just accept a lack of visibility. The query is even more difficult if you want it to be resilient to changes in the configuration.
3. How much data is my receiver emitting? Same convoluted answer as above.
4. In a linear pipeline, what is the net effect of all my processors? Same convoluted answer as above (vs just comparing outgoing from the receiver to incoming of the exporter).

The same problems exists if you only measure outgoing. The problem are even worse if we actually pin metrics to edges by including attributes from both components.

Reply:
> It sounds to me like these issues with having one metric per edge are caused by the fan-in and fan-out parts of the pipeline being "implicit".

Edges are implicit in our configuration, but components are explicit. Similarly, even in the code edges are barely more than a function call between components, whereas components are actual instances with state and attributes. I think this reflects the configuration, documentation, etc, and therefore it makes sense to ascribe signals to components rather than edges.

Reply:
I'm resolving this comment, but please reopen if you think the justifications are not sufficient.

Reply:
I get the **philosophical** argument that components are easier to notice or to reason about. I generally tend to favor solutions that are easy to reason about even if not completely optimized.

In this case, I'm on the fence. I'd argue that **conceptually** the information we want to express - "how much telemetry has been sent from component A to B" -  is not a property of either of the components (either A or B), but rather it's a property of the relation between those components.